/*
 * LICENSE HERE
 */

syntax = "proto3";

import "common.proto";

package bareos.config;

service Config {
  rpc ListClients (ListClientsRequest) returns (ListClientsResponse);
  rpc ListJobs (ListJobsRequest) returns (ListJobsResponse);
  rpc ListCatalogs (ListCatalogsRequest) returns (ListCatalogsResponse);


  rpc GetDefinition (GetDefinitionRequest) returns (GetDefinitionResponse);
};

message ListClientsRequest {
};

message ClientId {
  bytes name = 1;
};
message Client {
  ClientId id = 1;
  string name = 2;
  string address = 3;
};

message ListClientsResponse {
  repeated Client clients = 1;
};

message JobId {
  bytes name = 1;
};

message Job {
  JobId id = 1;
  string name = 2;
  bareos.common.JobType type = 3;

  string description = 4;
};

message JobTypeFilter {
  bareos.common.JobType select = 1;
};

message JobFilter {
  oneof filter_type {
    JobTypeFilter type = 1;
  }
};

message ListJobsRequest {
  // if no filter is set, then all jobs are accepted,
  // otherwise only jobs are accepted that match all
  // of the supplied filters
  repeated JobFilter filters = 1;
};

message ListJobsResponse {
  repeated Job jobs = 1;
};

message CatalogId {
  bytes name = 1;
};

message Catalog {
  CatalogId id = 1;
  string name = 2;
  string dbname = 3;
};

message ListCatalogsRequest {
};

message ListCatalogsResponse {
  repeated Catalog catalogs = 1;
};

message GetDefinitionRequest {
  oneof config_type {
    ClientId client = 1;
    CatalogId catalog = 2;
    JobId job = 3;
  };
};

message GetDefinitionResponse {
  map<string, string> set_options = 1;
};
